Global styles: add revisions UI

ramonjd opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

Related to:

- https://github.com/WordPress/gutenberg/issues/49601 (parent issue)
- https://github.com/WordPress/gutenberg/issues/41232
- https://github.com/WordPress/gutenberg/issues/33094


## What?
This PR displays revisions to the global styles in the global styles sidebar (revisions for the wp_global_styles custom post type) in a timeline-like list.

At the moment is a limit imposed by the rest controller of 100 revisions.

The revisions panel will only be available once there are more than two revisions. This is because the theme global styles default state is not treated as a revision. It's [possible to add it in though](https://github.com/WordPress/gutenberg/pull/50089#issuecomment-1532324313). 

## How?
Building on the work in :

- https://github.com/WordPress/gutenberg/pull/49973
- https://github.com/WordPress/gutenberg/pull/49974

## TODO
- [x] Maybe extract components in screen-revisions.js into separate files
- [x] Extract changes that are testable on their own, e.g., [API changes](https://github.com/WordPress/gutenberg/pull/50234)
- [x] Add E2E tests
- [x] Whatever else swings by

## Testing Instructions
1. Make a few changes to your site's global styles, save each one.
2. Open the revisions view via the global styles context menu
3. Check that the "currently-saved" revision is correct. Now select a revision and ensure you can preview it
4. Click on the "Load revision" to load it in the editor and close the revisions view
5. Save the post
6. Make some changes to global styles but don't save. Changes may include global style properties such as color, or layout and also block-level styles or custom color palettes. The more the better. Go for it!
7. Open the revisions view and check that there is an "Unsaved changes" item at the start of the list. Now try to load a revision as per steps 2-4
8. You should be prompted to exit and save your current changes from step 6, or go ahead and load the revision into the editor
9. Now open the revisions view
10. Play around with the panel and make sure it closes as expected, e.g., when switching to code editor view or opening the stylebook

Run the E2E tests if you want
```
npm run test:e2e:playwright -- --headed site-editor/user-global-styles-revisions.spec
```

## Screenshots or screencast


https://user-images.githubusercontent.com/6458278/236083659-ec8e1e53-dc90-43d0-9701-7d5d07269965.mp4

ramonjd committed:
initial commit.

ramonjd committed:
Adding state for revisions
Adding revisions global styles sidebar UI
Adding revisions fill

Adding revisions components and styles.

Added e2e tests

mtias commented:
Great progress!

@jasmussen @jameskoster @SaxonF — one thing we should be getting ahead of is treating "styles" as a more comprehensive section where revisions, style book, etc, can feel more integrated. Thinking of the ways in which we present the header and header tools, for example.

jameskoster commented:
Thanks @ramonjd. Please give me a ping when the general UX is closer to the latest in https://github.com/WordPress/gutenberg/issues/49601, and I'm happy to help polish the pixels!

ramonjd commented on lib/experimental/class-gutenberg-rest-global-styles-revisions-controller.php:
```diff
@@ -135,24 +135,24 @@ public function prepare_item_for_response( $item, $request ) {
 
 		if ( rest_is_field_included( 'author', $fields ) ) {
-			$data['author'] = (int) $parent->post_author;
+			$data['author'] = (int) $item->post_author;
```
I have a separate PR for these changes including unit tests: https://github.com/WordPress/gutenberg/pull/50117

Just committing them here to aid testing

ramonjd commented:
> Please give me a ping when the general UX is closer to the latest in https://github.com/WordPress/gutenberg/issues/49601, and I'm happy to help polish the pixels!

👋 @jameskoster I've got this branch working according to the [last version of my explorations](https://github.com/WordPress/gutenberg/pull/46667#issuecomment-1512529945).

The way I've set it up means that we can load and preview revisions without affecting the editor state and, therefore, any unsaved changes. 

Just wanted to check how far we could take things according to the [timeline demo you posted.](https://github.com/WordPress/gutenberg/issues/49601#issuecomment-1518004558)

I like the time line look, and I'd like to take a stab at it. I just had a couple of thoughts:

❓ Would the number of revisions impact the layout? What if there are > 50? I'm loading them in a select dropdown for now when the number of revisions is > 10, which is practical, but doesn't look fabulous.

<img width="273" alt="Screenshot 2023-04-27 at 3 40 01 pm" src="https://user-images.githubusercontent.com/6458278/234769883-329b49ee-3930-4d01-8382-d480b5be7981.png">

❓ Is the save modal still warranted?  It might be a tricky where there are template and global styles changes. I assume we'd only want to save global styles and not template changes. I think it'd be beneficial to do that in a follow up PR given the complexity, for ease of review.  

I've already [mentioned it](https://github.com/WordPress/gutenberg/pull/46667#issuecomment-1514188769), I believe we don't need to introduce a new save flow now that previewing revisions is independent of the editor. Users can open revisions, try them out, and return to the editor in whatever state they left it in, that is, any unsaved changes are unaffected. 

Thanks!

jameskoster commented:
> The way I've set it up means that we can load and preview revisions without affecting the editor state and, therefore, any unsaved changes

Oh, nice, that is excellent news! That probably means we don't need the save modal, but may need a dedicated 'Load this revision' button which my designs didn't have. I'll revisit my prototype and share an update soon. 

> Would the number of revisions impact the layout?

I don't _think_ so. Any reason not to just allow the timeline to scroll?

jameskoster commented:
Storing unsaved changes makes for a much simpler design :) 


https://user-images.githubusercontent.com/846565/234862626-399fa1fa-6a0b-4a00-b8fc-b1450807a2bb.mp4


* The current state, whether it is an actual revision or just unsaved changes appears active at the top of the timeline.
* Selecting any other revision reveals the 'Load revision' button.
* Confirmation dialog informs the user that unsaved changes will be lost.
* On load, any existing unsaved changes are replaced with the revision.
* Nothing is saved until the user clicks the Save button, this is when a new revision is created.

Until Styles exists as a more comprehensive section as [mentioned above](https://github.com/WordPress/gutenberg/pull/50089#issuecomment-1523201779), we'll need to think about what happens when a user selects a revision, doesn't load it, then exits the Revisions timeline somehow (clicks back, closes Styles panel, opens Inspector, etc). In other words, what happens if you click the `<` button here:

<img width="1280" alt="Screenshot 2023-04-27 at 13 27 56" src="https://user-images.githubusercontent.com/846565/234861997-3e5f5812-f71a-4710-aa4d-20effb1a4b78.png">

I think it's probably fine to just return to the stored unsaved changes for now, but it feels like a situation where it could be good to get confirmation from the user.

kevin940726 commented on test/e2e/specs/site-editor/user-global-styles-revisions.spec.js:
```diff
@@ -0,0 +1,124 @@
+
+test.use( {
+	editor: async ( { page }, use ) => {
+		await use( new Editor( { page, hasIframe: true } ) );
```
The `hasIframe` property has been removed 🙂 .

kevin940726 commented on test/e2e/specs/site-editor/user-global-styles-revisions.spec.js:
```diff
@@ -0,0 +1,124 @@
+		userGlobalStylesRevisions,
+	} ) => {
+		// Navigates to Styles -> Typography -> Text and click on a size.
+		await page.getByRole( 'button', { name: 'Styles' } ).click();
```
Tip: We can chain locators to write more readable queries so that we don't necessarily need some of the comments:

```suggestion
		await page
			.getByRole( 'region', { name: 'Editor top bar' } )
			.getByRole( 'button', { name: 'Styles' } )
			.click();
```

kevin940726 commented on test/e2e/specs/site-editor/user-global-styles-revisions.spec.js:
```diff
@@ -0,0 +1,124 @@
+			await expect( revisionButtons ).toHaveCount(
+				currentRevisions.length + 2
+			);
+		} else {
```
Nit: To retain the same test flow, let's remove the `else` statement and just perform the rest of the test unconditionally.

kevin940726 commented on test/e2e/specs/site-editor/user-global-styles-revisions.spec.js:
```diff
@@ -0,0 +1,124 @@
+				.getByRole( 'button', { name: 'Typography Text styles' } )
+				.click();
+			await page.click(
+				'role=radiogroup[name="Font size"i] >> role=radio[name="Large"i]'
```
Nit: Why the sudden change of the locator style 😆 ?

ramonjd replied:
To make life interesting of course! 😄 
I'll make it consistent according to the chaining tip above. Thank you!

ramonjd committed:
Making selectors more consistent in e2e tests

ramonjd commented:
> Any reason not to just allow the timeline to scroll?

You know what, I don't know! 😄

> Storing unsaved changes makes for a much simpler design :)

Thanks a lot for taking the time to revisit the design. 

 ~~I'll throw something up so we can play around with it.~~ 

@jameskoster I've made pretty decent headway on the design today. See the updated GIF in the PR description. 

Code-wise there's a bit to clean up, including splitting out API and other changes that might require separate reviews. 

Also, ignore the pesky CI test fails... I'll deal with them later 😜

ramonjd committed:
This commit:
- redesigns the revisions list to show a timeline view
- changes the revisions endpoint response (and tests) to return a human friendly diff only. The frontend can format dates on the fly
- Adds a slot to the edit side sidebar that can position a component at the bottom of the interface sidebar
- updates e2e tests

ramonjd committed:
This commit:
- adds wordpress/date to package-lock.jdon
- ensures that we correctly modify the user revisions to add unsaved and islatest flags

jasmussen commented:
Following the discussion here, the work looks amazing!

I want to note for consideration than in addition to the completely visual diffing here, we probably still want a way to see some indication of changes in the canvas itself, highlights or otherwise. We use a diffing tool for blocks that can't be parsed:

<img width="892" alt="Screenshot 2023-04-28 at 09 43 08" src="https://user-images.githubusercontent.com/1204802/235087375-6a98baed-fe9f-4d14-828f-03bd7cd18fa1.png">

In the vein of that, can we use `ins` and `del` tags to do the same in the canvas, when you compare the latest with a previous version?

<img width="720" alt="Diffing" src="https://user-images.githubusercontent.com/1204802/235087540-1a5ca361-39cd-4a90-b018-a3083bc49b4e.png">

Edit: not speaking this branch, but I just used Figma's version history tool today which is visual only, with no indication or pointers as to what changed, and just noting we need to do better than that.

jameskoster commented:
Thanks @ramonjd. With the bones in place I can hopefully work on some of the styling / details next week. Watch this space!

@jasmussen For styles a side-by-side could work:

https://user-images.githubusercontent.com/846565/235176038-57ec26b4-1687-45bf-a0ae-f446711df207.mp4

But it doesn't seem essential for a v1? It would be nice to have as an option.

mtias commented:
Side by side is nice, we could also have a mode that does the before / after juxtaposed where you drag a line in the middle to compare. Agreed not essential for v1.

jawomg commented:
[inter-noise-2009-paper-on-changes-in-new-editions-of-ecma-standards.pdf](https://github.com/WordPress/gutenberg/files/11360196/inter-noise-2009-paper-on-changes-in-new-editions-of-ecma-standards.pdf

jawomg commented:
#50134

ramonjd committed:
This commit:
- extends the e2e test a little bit and adds a TODO to do it better
- Adds ally labels to the color panel

ramonjd committed:
This commit:
- uses a clientside human-readable time diff method instead of the server side one. This is to make the response as close to the post revisions response as possible (for now)
- makes the current timeline dot blue

This commit:

ramonjd commented on packages/date/src/index.js:
```diff
@@ -581,6 +581,20 @@ export function getDate( dateString ) {
+ *
+ * @return {string} Human-readable time difference.
+ */
+export function humanTimeDiff( from, to ) {
```
This is nicked from https://github.com/WordPress/gutenberg/pull/49980

ramonjd commented:
>  But it doesn't seem essential for a v1? It would be nice to have as an option.

Just to illustrate, this took me about 10 mins to put together on top of this PR:


https://user-images.githubusercontent.com/6458278/235428869-668a38f6-97f9-4ea1-b078-a3a6da103cc8.mp4



So I agree, I expect performance/mobile/other issues to accompany this exploration so it might be best put on the boil until we nail a v1. 

Cheers!!

jasmussen commented:
Oh nice, keep that in a branch somewhere even if it doesn't land in the V1. For a split view and indeed the revisions container, it seems important to ensure the code is in a good place. Much of this is stuff that could benefit from being animated in place; everyting from "zooming out" when entering the revisions view, or even _the older version_ appearing from underneath the current version when it initially appears. Making sure we can orchestrate things like this will be good. Speaking of exploring split view, should it scroll both containers at the same time? I've seen that in some classic diffing tools.

ramonjd commented:
> Speaking of exploring split view, should it scroll both containers at the same time?

I think so. I was about to add it to the demo, but someone rang the dinner bell 🥫

ramonjd committed:
This commit:
- removes the specific author/user properties from the rest controller
- grabs revision author information in the JS
- shows a loading wheel while we grab revisions and user data

jameskoster commented:
Nice, it's good to know the split view shouldn't be too much extra work on the technical side. I'd +1 keeping it separate for now as it probably needs a dedicated design exploration to get right.

jameskoster committed:
Label titles

jameskoster committed:
Style adjustments

Adding a reset to theme default item to the revisions list.

Adding a dependencies to the `useMemo` so that it knows to update when the number of revisions change.

This is so we can ensure we add the right metadata in the right order, e.g., `isLatest`

Extracted hooks and components from screen revisions monster file

jameskoster commented:
@ramonjd I noticed I cant open the revisions timeline when there's only one revision. I'm not sure if there's any logic to that but I found it a bit confusing?

I just pushed some style changes. I think I might have broken the gravatars though I can't see how. Would you mind taking a look? 🙏

ramonjd commented:
> I noticed I cant open the revisions timeline when there's only one revision. I'm not sure if there's any logic to that but I found it a bit confusing?

It's because the initial global style custom post isn't considered a "revision" . So if we reduce the condition to `const hasRevisions = revisionsCount >= 1;`, there's nothing with which we can compare it. It's the same for the [templates](https://github.com/WordPress/gutenberg/blob/a1f736bb0bda91e8637bd6fa7be89bdf5d7605a1/packages/edit-site/src/components/sidebar-edit-mode/template-card/last-revision.js#L35) at the moment.

<img width="500" alt="Screenshot 2023-05-03 at 9 16 27 am" src="https://user-images.githubusercontent.com/6458278/235805864-0f62acf7-43ec-44df-b2eb-20a0e27d8bd7.png">

Maybe we could inject an item at the bottom of the list, something like "reset to default" or a better label, that will load the reset state.

<img width="313" alt="Screenshot 2023-05-03 at 9 25 58 am" src="https://user-images.githubusercontent.com/6458278/235806590-4ef132a1-ccd4-4837-a7c6-2858f6968fd7.png">


I'll see if it's possible.


> I just pushed some style changes. I think I might have broken the gravatars though I can't see how. Would you mind taking a look? 🙏

The gravatar size options don't include a `'32'`. I'll revert back to `'24'` for now.

ramonjd commented on lib/experimental/class-gutenberg-rest-global-styles-revisions-controller.php:
```diff
@@ -265,11 +265,16 @@ public function get_item_permissions_check( $request ) {
+		/*
+		 * The same check as WP_REST_Global_Styles_Controller->get_theme_items_permissions_check.
+		 */
+		if ( ! current_user_can( 'edit_theme_options' ) ) {
```
I'll pull this change into a separate PR with tests as well.

ramonjd replied:
#50270

ramonjd commented:
> Maybe we could inject an item at the bottom of the list, something like "reset to default" or a better label, that will load the reset state.

I've got this so far. Is it useful? (I just need to fix the `is-current` state of the buttons)

Basically the revisions panel is enabled after 1 x save, where we'll display the current revision and the reset state.

https://user-images.githubusercontent.com/6458278/235814784-5c3d64cf-1169-458d-aea7-55150d7cc07d.mp4

ramonjd committed:
Adding tests for the useGetGlobalStylesRevisions hook

ramonjd committed:
Extracting SCSS to external file
Renaming date > meta in classname of DIV that wraps revision date and author meta
Extra checks so that we don't render the meta block at all if there's no data

jameskoster commented:
@ramonjd Ah sorry for opening that can of worms. We can probably leave it for now since we already have the 'reset to defaults' action in the ellipsis menu. Perhaps we can consolidate that with an initial 'default' revision like in your demo, but I'd suggest we explore that in a follow-up.

> The gravatar size options don't include a '32'. I'll revert back to '24' for now.

Ah, ok. Since the image renders at 16x16 the source needs to be at least double that size to avoid any pixelation on 'retina' displays. Would 48 work?

ramonjd commented:
> We can probably leave it for now since we already have the 'reset to defaults' action in the ellipsis menu. Perhaps we can consolidate that with an initial 'default' revision like in your demo, but I'd suggest we explore that in a follow-up.

No worries. At least we know it's possible now. I'll keep it stashed in a revision. 👍 

> Ah, ok. Since the image renders at 16x16 the source needs to be at least double that size to avoid any pixelation on 'retina' displays. Would 48 work?

~~Should be fine. Maybe, if it suits the design better we could go 48 and scale it to 32 via CSS? I'll play around.~~ 

I'll add it in. Should work I reckon.

Cheers!

ramonjd committed:
Reverting the reset theme default button and ensuring that revisions panel only display when there are more than two revisions.
Updated tests accordingly.

bumping gravatar size to 48

Updating other E2E tests after having added aria labels to the color controls

Update e2e button spec after changes to button labels

ramonjd commented on packages/block-editor/src/components/global-styles/color-panel.js:
```diff
@@ -230,6 +230,11 @@ function ColorPanelDropdown( {
 							{ 'is-open': isOpen }
 						),
 						'aria-expanded': isOpen,
+						'aria-label': sprintf(
```
Extracting out this change plus necessary updates to E2E tests over in https://github.com/WordPress/gutenberg/pull/50307 to reduce a bit of clutter

jameskoster commented:
> Should be fine. Maybe, if it suits the design better we could go 48 and scale it to 32 via CSS? I'll play around.

_Displaying_ at 16px is important because that's the same space occupied by the accompanying text label. We just need to ensure the source image is at least 2*16. So if 48 work that'll do the trick!

jasmussen commented:
Thanks both for all the work you're doing here!

ramonjd commented:
> Displaying at 16px is important because that's the same space occupied by the accompanying text label. We just need to ensure the source image is at least 2*16. So if 48 work that'll do the trick!

Excellent! I misread your comment (got to stop skim reading), so I just bumped it to 48. 👍 Looks great.

noisysocks commented:
My notes from testing:

- Nice! This is awesome!!

- For the _Revisions_ link that appears in the ellipsis menu, could we borrow @javierarce's design that I implemented (but never shipped) in https://github.com/WordPress/gutenberg/pull/42175? I think it communicates better that the number is dynamic.

   ![159279715-18cb656b-9eaa-4cbe-ab0f-bd099b0f23d6](https://user-images.githubusercontent.com/612155/236356806-c3c0f977-a8ac-473b-b1e5-c2f98eec93c5.png)

- It would be great if we could label the revisions more intelligently than saying _Changes saved_ repeatedly. For example if all of the changes were typography changes then it could say _Typography changes_. I don't know, maybe that's too hard.

- The word _Load_ in the _Load revision_ button feels too technical to me. Could it be _Apply_? Also on tall viewports the button appears so far away from the revisions that it feels disconnected. 

- The copy in this confirmation prompt is hard to grok with a single glance. The question (_Do you want to...?_) should be in the title and the buttons should be labeled with what they'll do. Have a look at the macOS confirmation you get when closing an unsaved document for inspiration.

   <img width="593" alt="Screenshot 2023-05-05 at 10 35 45" src="https://user-images.githubusercontent.com/612155/236357252-54e0d900-8b8e-4014-859a-d1e3e91f0cea.png">

   I'd suggest something like this:

   ```
   Do you want to replace your unsaved changes in the editor?

   There are currently unsaved style changes. Loading this revision will discard these changes.

   [ Cancel ] [ Discard unsaved changes ]
   ```

noisysocks commented on packages/block-editor/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -1106,8 +1106,11 @@ const processCSSNesting = ( css, blockSelector ) => {
-export function useGlobalStylesOutput() {
+export function useGlobalStylesOutput( customMergedConfig = null ) {
 	let { merged: mergedConfig } = useContext( GlobalStylesContext );
+	if ( !! customMergedConfig ) {
```
The `!!` is unnecessary in an `if`.

```suggestion
	if ( customMergedConfig ) {
```

noisysocks commented on packages/block-editor/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -1106,8 +1106,11 @@ const processCSSNesting = ( css, blockSelector ) => {
 };
 
-export function useGlobalStylesOutput() {
+export function useGlobalStylesOutput( customMergedConfig = null ) {
```
Instead of adding an "override" param here could we split `useGlobalStylesOutput` into two hooks? One that loads the merged config from context and output sit, and another that just outputs what is given to it. I think that would be a little clearer to grok, and it means we're not accessing context unnecessarily when a `customMergedConfig` is provided.

ramonjd replied:
This is a good idea. 🍺 Implementing...

noisysocks commented on packages/block-editor/src/components/global-styles/utils.js:
```diff
@@ -376,3 +377,29 @@ export function scopeSelector( scope, selector ) {
+ *
+ * @return {boolean} Whether `original` and `variation` match.
+ */
+export function isGlobalStyleConfigEqual( original, variation ) {
```
nit: My English teacher probably would be happier with `areGlobalStyleConfigsEqual`.

ramonjd replied:
A+

noisysocks commented on packages/block-editor/src/components/global-styles/utils.js:
```diff
@@ -376,3 +377,29 @@ export function scopeSelector( scope, selector ) {
+ * @return {boolean} Whether `original` and `variation` match.
+ */
+export function isGlobalStyleConfigEqual( original, variation ) {
+	if ( ! original || ! variation ) {
```
Shouldn't `isGlobalStyleConfigEqual( null, null )` be true? I guess it doesn't matter in practice 😅

ramonjd replied:
Sounds better logically. I'll do a type check or something so it handles `( null, null )` and `( (), undefined )` etc as expected.

Cheers!

noisysocks commented on packages/block-editor/src/components/global-styles/utils.js:
```diff
@@ -376,3 +377,29 @@ export function scopeSelector( scope, selector ) {
+	return (
+		fastDeepEqual( original?.styles, variation?.styles ) &&
+		fastDeepEqual( original?.settings, variation?.settings )
+	);
```
Why have this function instead of just using `fastDeepEqual`? For semantic reasons? (Not against that.)

ramonjd replied:
The functionality is used in a couple of places (style variations) and given the specific global styles properties we're checking I wanted to make something reusable.

noisysocks commented on packages/core-data/src/actions.js:
```diff
@@ -210,6 +210,28 @@ export function receiveThemeSupports() {
+ *
+ * @return {Object} Action object.
+ */
+export function __experimentalReceiveThemeGlobalStyleRevisions(
```
We should use a private action instead of the `__experimental` prefix which is a little meaningless now that we have to support experimental functions that are in Core.

Or can just mark it as stable. It seems a well established pattern that I doubt will change.

ramonjd replied:
I did try to make it private at the start, but ran into issues with the selector's resolver. 

I can mark it as stable, it's a pattern I copied as you say 😄

noisysocks commented on packages/core-data/src/resolvers.js:
```diff
@@ -500,6 +500,50 @@ export const __experimentalGetCurrentThemeGlobalStylesVariations =
 		);
 	};
 
+export const __experimentalGetCurrentThemeGlobalStylesRevisions =
```
Missing doc comment.

Also probably doesn't need to be `__experimental`. In all likelihood it won't change and will end up as yet another item on our experimental API audit list for a dozen releases in a row.

noisysocks commented on packages/edit-site/src/components/editor-canvas-container/index.js:
```diff
@@ -46,6 +48,7 @@ const {
 
 function EditorCanvasContainer( {
 	children,
+	showCloseButton = false,
```
How about instead of a new prop we don't show the button if `onClose` is not provided?

noisysocks commented on packages/edit-site/src/components/global-styles/screen-revisions/index.js:
```diff
@@ -0,0 +1,155 @@
+				) }
+			</div>
+			{ isLoadingRevisionWithUnsavedChanges && (
+				<RestoreGlobalStylesRevisionModal
```
We should use the existing `ConfirmDialog` component instead of our own custom modal.

https://wordpress.github.io/gutenberg/?path=/story/components-experimental-confirmdialog--default

ramonjd replied:
**TIL** Much easier, thanks!

noisysocks commented on packages/edit-site/src/components/global-styles/screen-revisions/restore-global-styles-revision-modal.js:
Should be able to delete this whole file if using `ConfirmDialog` in `ScreenRevisions`.

noisysocks commented on packages/edit-site/src/components/global-styles/screen-revisions/revisions-buttons.js:
```diff
@@ -0,0 +1,129 @@
+	if ( isUnsaved ) {
+		return sprintf(
+			/* translators: %(name)s author display name */
+			__( 'Unsaved changes by %(name)s' ),
```
Whoa I didn't know you could name the `sprintf` params! Amazing 🤯

noisysocks commented on packages/edit-site/src/components/global-styles/screen-revisions/revisions-buttons.js:
```diff
@@ -0,0 +1,129 @@
+	const date = getDate( revision?.modified );
+	const formattedDate = dateI18n(
+		// translators: a compact version of the revision's modified date.
+		_x( 'j M @ H:i', 'formatted version of revision last modified date' ),
```
I'd use `at` instead of `@` since this is going into an `aria-label` and screen readers / translators might struggle with the symbol.

Also, the default locale (which is then translated) is `en_US` so we should use the American way of writing dates and times here.

```suggestion
		_x( 'M j g:i a', 'formatted version of revision last modified date' ),
```

noisysocks replied:
We could also just use `wp.date.getSettings().formats.datetimeAbbreviated` and not reinvent the wheel. That way it's somebody else's problem 😀

ramonjd replied:
> i'd use at instead of @ since this is going into an aria-label and screen readers / translators might struggle with the symbol.

It's the format that we use on the revisions pages, that's why I went with it.

<img width="342" alt="Screenshot 2023-05-05 at 12 28 01 pm" src="https://user-images.githubusercontent.com/6458278/236366104-d59cdaf8-5411-4394-ab96-49ba4b7bff7d.png">

> We could also just use wp.date.getSettings().formats.datetimeAbbreviated and not reinvent the wheel. That way it's somebody else's problem 

That sounds like a much better idea 😄

noisysocks commented on packages/edit-site/src/components/global-styles/screen-revisions/revisions-buttons.js:
```diff
@@ -0,0 +1,129 @@
+								'is-current': isActive,
+							}
+						) }
+						key={ `user-styles-revision-${ id }` }
```
These are only used by React so don't need to give it anything beyond a unique string.

```suggestion
						key={ id }
```

noisysocks commented on packages/edit-site/src/components/global-styles/screen-revisions/revisions-buttons.js:
```diff
@@ -0,0 +1,129 @@
+		<ol
+			className="edit-site-global-styles-screen-revisions__revisions-list"
+			aria-label={ __( 'Global styles revisions' ) }
+			role="group"
```
Doesn't setting the role to `group` negate the point of using an `ol`?

ramonjd replied:
This was originally a naïve cheat code for making it easier to target in e2e tests :trollface: Removing...

noisysocks commented on packages/edit-site/src/components/global-styles/screen-revisions/revisions-buttons.js:
```diff
@@ -0,0 +1,129 @@
+							onClick={ () => {
+								onChange( revision );
+							} }
+							aria-label={ getRevisionLabel( revision ) }
```
Generally one should be able to hover over a relative date and see the actual date, so I'd switch this from `aria-label` to `label`.

noisysocks commented on packages/edit-site/src/components/global-styles/screen-revisions/revisions-buttons.js:
```diff
@@ -0,0 +1,129 @@
+		>
+			{ userRevisions.map( ( revision ) => {
+				const { id, author, isLatest, modified } = revision;
+				const authorAvatar = author?.avatar_urls?.[ '48' ];
```
Remind me what `48` is again? Will it always exist in the object?

ramonjd replied:
https://developer.wordpress.org/reference/functions/rest_get_avatar_sizes/

There's no filter I can find so we'll always get `24, 48, 96`

noisysocks commented on packages/edit-site/src/components/global-styles/screen-revisions/use-get-global-styles-revisions.js:
```diff
@@ -0,0 +1,107 @@
+};
+
+const { GlobalStylesContext } = unlock( blockEditorPrivateApis );
+export default function useGetGlobalStylesRevisions() {
```
Nit but `useGetGlobalStylesRevisions` makes me think the hook will return a function named `getGlobalStylesRevisions`, which it does not. I'd name it `useGlobalStylesRevisions`.

noisysocks commented on packages/edit-site/src/components/global-styles/screen-revisions/use-get-global-styles-revisions.js:
```diff
@@ -0,0 +1,107 @@
+		 * new "revision" item that represents the unsaved changes.
+		 */
+		_modifiedRevisions = revisions.map( ( revision, index, _array ) => {
+			if ( 0 === index && _array[ index ]?.id !== 'unsaved' ) {
```
Probably simpler to do this outside of the `map`?

```js
		_modifiedRevisions = revisions.map( ( revision ) => {
			return {
				...revision,
				author: authors.find(
					( author ) => author.id === revision.author
				),
			};
		} );

		if ( _modifiedRevisions[ 0 ]?.id !== 'unsaved' ) {
			_modifiedRevisions[ 0 ].isLatest = true;
		}
```

noisysocks commented on packages/edit-site/src/components/global-styles/screen-revisions/use-get-global-styles-revisions.js:
```diff
@@ -0,0 +1,107 @@
+			return {
+				...revision,
+				author: {
+					...authors.find(
```
This `...` isn't necessary.

noisysocks commented on packages/edit-site/src/components/global-styles/screen-revisions/use-get-global-styles-revisions.js:
```diff
@@ -0,0 +1,107 @@
+			};
+		} );
+
+		if ( isDirty && ! isEmpty( userConfig ) && !! currentUser ) {
```
Don't need `!!` in an `if`.

```suggestion
		if ( isDirty && ! isEmpty( userConfig ) && currentUser ) {
```

noisysocks commented on packages/edit-site/src/components/global-styles/screen-revisions/use-get-global-styles-revisions.js:
```diff
@@ -0,0 +1,107 @@
+				modified: new Date(),
+			};
+
+			_modifiedRevisions = [ unsavedRevision ].concat(
+				_modifiedRevisions
+			);
```
Could use `unshift` to prepend an item here.

```suggestion
			_modifiedRevisions.unshift( unsavedRevision );
```

ramonjd replied:
~~That was deliberate to avoid mutation. Before I created the custom hook I ran into problems with that, but now it might not be so important. I'll check.~~

Withdraw that, I was thinking of something else.

ramonjd commented:
Thanks for testing @noisysocks 

> For the Revisions link that appears in the ellipsis menu, could we borrow @javierarce's design that I implemented (but never shipped) in https://github.com/WordPress/gutenberg/pull/42175? I think it communicates better that the number is dynamic.

Great stuff. I'll see what I can pilfer from your PR! 😄 

> It would be great if we could label the revisions more intelligently than saying Changes saved repeatedly. For example if all of the changes were typography changes then it could say Typography changes. I don't know, maybe that's too hard.

I'm tending to agree about the repetition - that they are changes, and they are saved, are both implied. I think I prefer to highlight the date of the change more than anything, which is the version I had before. 

Maybe we could even take inspiration from Google docs and group by month, with collapsible sections.

 It wouldn't be a stretch to check the top-level styles/blocks and display an "extract" of the changes either.

<img width="300" src="https://user-images.githubusercontent.com/6458278/236362936-f8d441a2-875f-4947-942b-01d50b6495b5.jpg" />


To keep things trucking along, and to avoid making this PR too big, I'd consider these, however, immediate follow up explorations.

noisysocks commented on packages/edit-site/src/components/sidebar-edit-mode/sidebar-fixed-bottom.js:
```diff
@@ -0,0 +1,26 @@
+	);
+}
+
+export { SidebarFixedBottomSlot };
```
No need to export it again.

ramonjd replied:
Yes because one's a slot, the other is the fill wrapper. 

Maybe I should rename `function SidebarFixedBottom` to `function SidebarFixedBottomFill` to make it clearer...

noisysocks reviewed (commented):
Really nice work here!

noisysocks commented:
> To keep things trucking along, and to avoid making this PR too big, I'd consider these, however, immediate follow up explorations.

Yep this is totally fine as is for a v1.

ramonjd committed:
This commit:
- refactors useGlobalOutput by splitting out the build logic into a separate hook so that we can pass a custom config
- uses wp.date.getSettings().formats.datetimeAbbreviated for formatted date in the revisions buttons
- removes unnecessary role on the ordered list
- updated copy
- replacing modal component with the confirm dialog component
- minor code optimizations
- making the revisions store methods stable
- updating tests
- removing shouldShowClose button prop on the editor-canvas-container

update e2e

ramonjd commented:
@noisysocks Thank you for the detailed review and for making this PR better!! 🙇

jasmussen commented:
I won't have time today to jump into the code or a deeper review, but wanted to briefly comment on this piece:

<img width="300" src="https://user-images.githubusercontent.com/6458278/236362936-f8d441a2-875f-4947-942b-01d50b6495b5.jpg" />

I have a feeling this is a quick mockup exactly to vet the idea, and I would start without the collapsible sections. But you could add subheadings to group by month. In this style:

<img width="61" alt="Screenshot 2023-05-05 at 09 10 29" src="https://user-images.githubusercontent.com/1204802/236396796-a0957241-1ab7-4212-9a5d-27333fbb9ac8.png">

If we find enough revisions in the future that it needs collapsible sections, we can add it at that time, I think.

ramonjd commented:
> I have a feeling this is a quick mockup exactly to vet the idea, and I would start without the collapsible sections. 

Yeah, it was a just a shoddy mockup to illustrate a point. 👍 I think this PR is in pretty decent shape. Happy to follow up as new ideas file in. Thanks!

jameskoster commented:
Good feedback there, thanks @noisysocks !

If it's easy to do I wouldn't be averse to switching the labels so that we get this:

<img width="267" alt="Screenshot 2023-05-05 at 10 06 18" src="https://user-images.githubusercontent.com/846565/236418439-a64b03a4-ec3e-49e2-bca4-05fa905d5941.png">

The details element seems more tricky and I'm not sure how useful that would be in practise. That one feels more like follow-up territory.

In terms of the entry-point, it might be good to reuse the UI for accessing post / template revisions:

<img width="245" alt="Screenshot 2023-05-05 at 10 02 29" src="https://user-images.githubusercontent.com/846565/236418803-dca7e641-c186-4d81-9fe9-9304c850428e.png">

ramonjd commented:
> If it's easy to do I wouldn't be averse to switching the labels so that we get this:

Great stuff, thanks @jameskoster - I'll take a shot at it. Shouldn't be too much effort to update the individual revision button layout this PR. 🙇 

> In terms of the entry-point, it might be good to reuse the UI for accessing post / template revisions:

Maybe for a follow up discussion? I think @mtias express a preference for it to live in the drop down initially.

Thank you!

ramonjd committed:
Tweaked the revisions buttons copy, shifting the date and meta around.
Updated tests

ramonjd commented:
> For the Revisions link that appears in the ellipsis menu, could we borrow @javierarce's design that I implemented (but never shipped) in https://github.com/WordPress/gutenberg/pull/42175? I think it communicates better that the number is dynamic.

Just to note that I tried to hammer this into the `<DropDown />` component this morning in `GlobalStylesActionMenu()`... I think it needs a bit of a refactor to make it work with the current list items. 

Also the menu would have to be slightly wider to accommodate the copy and icon.

The items in the controls prop don't render elements, so maybe I can move them all over to MenuItems or something. I haven't looked too deeply into alternatives, but I'll note as a follow up.

Thanks again for all the help!

noisysocks commented:
Yeah `GlobalStylesActionItem` should be change to pass `<MenuItem>`s via `children` instead of using the `controls` prop. While we're at it we should use `<MenuGroup>` to group the items into logical groups since there's several items now and it's getting to be a bit incoherent. Can do this in a follow-up though—this PR is big enough.

noisysocks reviewed (approved):
Great work on this!

ramonjd commented:
Documenting follow ups:
- Explore adding an [initial 'default' revision to the list](https://github.com/WordPress/gutenberg/pull/50089#issuecomment-1532324313)
- [Refactor GlobalStylesActionMenu to use MenuGroup/Item](https://github.com/WordPress/gutenberg/pull/50089#issuecomment-1537622415) - will enable alternative revision badge designs
- Experiment with [side-by-side view ](https://github.com/WordPress/gutenberg/pull/50089#issuecomment-1529396895)

Mamaduka commented:
Hi, @ramonjd

The E2E tests for this feature seem to be flaky -https://github.com/WordPress/gutenberg/issues?q=is%3Aissue+is%3Aopen+in%3Abody+user-global-styles-revisions.spec.js.

ramonjd commented:
Thanks @Mamaduka 

I'll check it out 👍

kevin940726 commented:
It seems like the issue is that we're not disabling the Styles welcome guide on the first load. So it always fails once but passes the next run. It's easy to miss, but there's an automatically-created issue in https://github.com/WordPress/gutenberg/pull/50089#issuecomment-1524693295. Also there were inline failure messages in the File tab too ([L60](https://github.com/WordPress/gutenberg/pull/50089/files#diff-29d5610a87cf1d6aa07b2e9620d908ad86b024a76756cc2de9c464332edf41f6R60) and [L117](https://github.com/WordPress/gutenberg/pull/50089/files#diff-29d5610a87cf1d6aa07b2e9620d908ad86b024a76756cc2de9c464332edf41f6R117)).

ramonjd commented:
> It seems like the issue is that we're not disabling the Styles welcome guide on the first load. 

Thanks @kevin940726!! Looks like I forgot to pass `skipWelcomeGuide = true` to `visitSiteEditor()`. I'll test it out

Edit: Oh, now I remember - I saw that `skipWelcomeGuide` was `true` [by default](https://github.com/WordPress/gutenberg/blob/de94c50198e78c551f978849b69024d82b349617/packages/e2e-test-utils-playwright/src/admin/visit-site-editor.ts#L30-L30). 

That's why I left it. 🤷 

Edit Edit: You said the styles welcome guide. I understood the other one, sorry.  🤦

ramonjd commented:
@kevin940726 @Mamaduka 

I have what I hope is a fix for the flakiness over at https://github.com/WordPress/gutenberg/pull/50454

Turns out I was leaving unsaved changes in the site editor so I think the test might have been trying to select an already-saved style. I don't know. But it appears to run more reliably locally now.

jameskoster commented:
@ramonjd one thing we missed here that would be good to address in a follow-up: Use the zoom-out view for the canvas, rather than the stylebook container.

ramonjd commented:
> one thing we missed here that would be good to address in a follow-up: Use the zoom-out view for the canvas, rather than the stylebook container.

Sounds like a good follow up. Thanks @jameskoster 👍 I'll make a note.
